crates/ra_hir/src/expr.rs

```diff
@@ -0,0 +1,740 @@
+use std::sync::Arc;
+
+use rustc_hash::FxHashMap;
+
+use ra_arena::{Arena, RawId, impl_arena_id};
+use ra_db::{LocalSyntaxPtr, Cancelable};
+use ra_syntax::ast::{self, AstNode, LoopBodyOwner, ArgListOwner, NameOwner};
+
+use crate::{Path, type_ref::{Mutability, TypeRef}, Name, HirDatabase, DefId, Def, name::AsName};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ExprId(RawId);
+impl_arena_id!(ExprId);
+
+/// The body of an item (function, const etc.).
+#[derive(Debug, Eq, PartialEq)]
+pub struct Body {
+    exprs: Arena<ExprId, Expr>,
+    pats: Arena<PatId, Pat>,
+    /// The patterns for the function's arguments. While the argument types are
+    /// part of the function signature, the patterns are not (they don't change
+    /// the external type of the function).
+    ///
+    /// If this `Body` is for the body of a constant, this will just be
+    /// empty.
+    args: Vec<PatId>,
+    /// The `ExprId` of the actual body expression.
+    body_expr: ExprId,
+}
+
+/// An item body together with the mapping from syntax nodes to HIR expression
+/// IDs. This is needed to go from e.g. a position in a file to the HIR
+/// expression containing it; but for type inference etc., we want to operate on
+/// a structure that is agnostic to the actual positions of expressions in the
+/// file, so that we don't recompute the type inference whenever some whitespace
+/// is typed.
+#[derive(Debug, Eq, PartialEq)]
+pub struct BodySyntaxMapping {
+    body: Arc<Body>,
+    expr_syntax_mapping: FxHashMap<LocalSyntaxPtr, ExprId>,
```

Comment:
Would it be pemature to extract IdBiMap here?

Reply:
It at least makes the naming easier :smile:

Reply:
Hm, there are a few cases where the mapping isn't perfectly bidirectional however -- e.g. with a `ParenExpr`, both it and the inner expr point to the same HIR expr. Which may not be a great thing either...

Reply:
Hm, makes sense. That's more arguments in favor of encapsulating this fuzziness into a separate type.